File: SLAM/YTImport.vb
Imports System.ComponentModel
Imports System.IO
Imports System.IO.File
Imports System.Text.RegularExpressions
Imports VideoLibrary

Public Class YTImport

    Public file As String
    Sub ProgressChangedHandler(sender, args)
        Console.WriteLine(args.ProgressPercentage)
    End Sub

    Private Sub ImportButton_Click(sender As Object, e As EventArgs) Handles ImportButton.Click
        Dim youtubeMatch As Match = New Regex("youtu(?:\.be|be\.com)/(?:.*v(?:/|=)|(?:.*/)?)([a-zA-Z0-9-_]+)").Match(TextBox1.Text)

        If youtubeMatch.Success Then
            PictureBox1.Visible = True
            TextBox1.Enabled = False
            ImportButton.Enabled = False
            DownloadWorker.RunWorkerAsync("youtube.com/watch?v=" & youtubeMatch.Groups(1).Value)
            ToolStripStatusLabel1.Text = "Status: Downloading"
        Else
            MessageBox.Show("Invalid YouTube URL.", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            TextBox1.Enabled = True
            ImportButton.Enabled = True
            PictureBox1.Visible = False
        End If
    End Sub

    Private Sub YTImport_KeyDown(sender As Object, e As KeyEventArgs) Handles MyBase.KeyDown
        Select Case e.KeyCode
            Case Keys.Enter
                If ImportButton.Enabled Then
                    ImportButton_Click(sender, Nothing)
                End If
            Case Keys.Escape
                DialogResult = Windows.Forms.DialogResult.Cancel
        End Select
    End Sub

    Private Sub DownloadWorker_DoWork(sender As Object, e As System.ComponentModel.DoWorkEventArgs) Handles DownloadWorker.DoWork
        Try

            If Not Directory.Exists(Path.GetFullPath("temp\")) Then
                Directory.CreateDirectory(Path.GetFullPath("temp\"))
            End If

            Dim video = YouTube.Default.GetVideo(e.Argument)
            Dim filename As String = String.Join("", video.FullName.Split(Path.GetInvalidFileNameChars()))
            Dim filepath = "temp\" & filename
            WriteAllBytes(filepath, video.GetBytes())
            e.Result = Path.GetFullPath(filepath)
        Catch ex As Exception
            Form1.LogError(ex)

            e.Result = ex
        End Try
    End Sub

    Private Sub DownloadWorker_ProgressChanged(sender As Object, e As ProgressChangedEventArgs) Handles DownloadWorker.ProgressChanged

    End Sub

    Private Sub DownloadWorker_RunWorkerCompleted(sender As Object, e As RunWorkerCompletedEventArgs) Handles DownloadWorker.RunWorkerCompleted
        Dim ExceptionType = e.Result.GetType()

        While ExceptionType IsNot Nothing
            If ExceptionType = GetType(Exception) Then
                MessageBox.Show(e.Result.Message & " See errorlog.txt for more info.", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
                TextBox1.Enabled = True
                ImportButton.Enabled = True
                Return
            End If

            ExceptionType = ExceptionType.BaseType
        End While

        file = e.Result
        DialogResult = Windows.Forms.DialogResult.OK
    End Sub

    Private Sub DonateLabel_Click(sender As Object, e As EventArgs) Handles DonateLabel.Click
        Process.Start("https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=RVLLPGWJUG6CY")
    End Sub

    Private Sub YTImport_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        PictureBox1.Visible = False
        TextBox1.Select()
        KeyPreview = True
    End Sub


    Private Sub PictureBox1_Click(sender As Object, e As EventArgs) Handles PictureBox1.Click
    End Sub
End Class

File: SLAM/YTImport.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using YoutubeExtractor;

namespace SLAM
{
    public partial class YTImport
    {
        public YTImport()
        {
            InitializeComponent();
            _ImportButton.Name = "ImportButton";
            _DonateLabel.Name = "DonateLabel";
        }

        public string file;

        public void ProgressChangedHandler(object sender, object args)
        {
            Console.WriteLine(args.ProgressPercentage);
        }

        private void ImportButton_Click(object sender, EventArgs e)
        {
            var youtubeMatch = new Regex(@"youtu(?:\.be|be\.com)/(?:.*v(?:/|=)|(?:.*/)?)([a-zA-Z0-9-_]+)").Match(TextBox1.Text);
            if (youtubeMatch.Success)
            {
                TextBox1.Enabled = false;
                ImportButton.Enabled = false;
                DownloadWorker.RunWorkerAsync("youtube.com/watch?v=" + youtubeMatch.Groups[1].Value);
                ToolStripStatusLabel1.Text = "Status: Downloading";
            }
            else
            {
                MessageBox.Show("Invalid YouTube URL.", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                TextBox1.Enabled = true;
                ImportButton.Enabled = true;
            }
        }

        private void DownloadWorker_DoWork(object sender, DoWorkEventArgs e)
        {
            try
            {
                IEnumerable<VideoInfo> videoInfos = DownloadUrlResolver.GetDownloadUrls(Conversions.ToString(e.Argument)).OrderBy(vid => vid.Resolution);
                var video = videoInfos.First(info => info.AdaptiveType == AdaptiveType.Audio && info.AudioType == AudioType.Aac || info.AdaptiveType == AdaptiveType.None && info.VideoType == VideoType.Mp4 && info.AudioBitrate >= 128);
                if (Information.IsNothing(video))
                {
                    if (videoInfos.Any(info => info.AdaptiveType == AdaptiveType.None && info.VideoType == VideoType.Mp4))
                    {
                        video = videoInfos.First(info => info.AdaptiveType == AdaptiveType.None && info.VideoType == VideoType.Mp4);
                    }
                    else
                    {
                        throw new Exception("Could not find download.");
                    }
                }

                if (video.RequiresDecryption)
                {
                    DownloadUrlResolver.DecryptDownloadUrl(video);
                }

                if (!Directory.Exists(Path.GetFullPath(@"temp\")))
                {
                    Directory.CreateDirectory(Path.GetFullPath(@"temp\"));
                }

                string filename = string.Join("", video.Title.Split(Path.GetInvalidFileNameChars()));
                var VideoDownloader = new VideoDownloader(video, Path.GetFullPath(@"temp\" + filename + video.VideoExtension));
                VideoDownloader.DownloadProgressChanged += (send, args) => DownloadWorker.ReportProgress(Convert.ToInt32(args.ProgressPercentage));
                VideoDownloader.Execute();
                e.Result = Path.GetFullPath(@"temp\" + filename + video.VideoExtension);
            }
            catch (Exception ex)
            {
                My.MyProject.Forms.Form1.LogError(ex);
                e.Result = ex;
            }
        }

        private void DownloadWorker_ProgressChanged(object sender, ProgressChangedEventArgs e)
        {
            ToolStripProgressBar1.Value = e.ProgressPercentage;
        }

        private void DownloadWorker_RunWorkerCompleted(object sender, RunWorkerCompletedEventArgs e)
        {
            if (e.Result.GetType() == typeof(Exception))
            {
                MessageBox.Show(Conversions.ToString(Operators.ConcatenateObject(e.Result.Message, " See errorlog.txt for more info.")), "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
            else
            {
                file = Conversions.ToString(e.Result);
                DialogResult = DialogResult.OK;
            }
        }

        private void DonateLabel_Click(object sender, EventArgs e)
        {
            Process.Start("https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=RVLLPGWJUG6CY");
        }

        private void YTImport_Load(object sender, EventArgs e)
        {
            TextBox1.Select();
        }
    }
}